src/Tools/dotnet-monitor/Egress/S3/MultiPartUploadStream.cs

```diff
@@ -47,13 +59,43 @@ public override async Task FlushAsync(CancellationToken cancellationToken)
 
     public async Task FinalizeAsync(CancellationToken cancellationToken)
     {
+        _finalize = true;
+
+        if (_writeSynchronousArtifacts != null)
+            await _writeSynchronousArtifacts;
+        
         if (Closed)
             throw new ObjectDisposedException(nameof(MultiPartUploadStream));
         if (_offset == 0)
             return;
         await DoWriteAsync(true, cancellationToken);
     }
 
+    public async Task WriteAsyncLoop(CancellationToken cancellationToken)
+    {
+        while (!_finalize || _syncTempBuffer.Count > 0)
```

Comment:
If something doesn't immediately call Write on the stream after creating it, this loop is likely to exit very quickly because nothing is in the buffer. I don't think this loop should depend on data being in the buffer.